File: AA/EXPERIMENTS/Amortized_aggregate.java
import java.util.*;

public class Amortized_aggregate {
    static int maxSize = 3, top = 0, pop_cost = 0, push_cost = 0, multipop_cost = 0;

    public static boolean isFull (){
        return top == maxSize-1;
    }

    public static void multipop (int k, int maxSize, Stack<Integer> st){
        for(int i = 0; i < k; i++) {
            if (!st.empty()) {
                int popped = st.pop();
                pop_cost++;
                System.out.println("Popped element "+popped+" stack = "+st);
            } else {
                break;
            }
        }
    }

    public static void main(String args[]) {
        Stack<Integer> stack = new Stack<>();

        int[] arr = {5,7,9,2,6,1,8,3};
        for(int i = 0; i < arr.length; i++) {
            if (arr[i] <= maxSize) {
                int k = arr[i];
                multipop(k,maxSize,stack);
                multipop_cost++;
            }
            stack.push(arr[i]);
            push_cost++;
            System.out.println("Pushed element "+i+" stack = "+stack);
            top++;
        }
        System.out.println("Cost of all operation = "+(push_cost+pop_cost));
        System.out.println("Cost of multipop operation = "+multipop_cost);
        System.out.println("Time complexity = O("+(push_cost+pop_cost-multipop_cost)/arr.length+")");
    }
}

File: AA/EXPERIMENTS/Amortized_aggregate.py
def push(element):
    global operations_count,unit,stack_len
    if len(stack)<=stack_len:
        stack.append(element)
        operations_count+=1
        unit.append(1)
        print(f"Stack after pushing {element} is {stack} and the unit array is {unit}")    
def pop():
    if len(stack)==0:
        return
    else:
        stack.pop()
def multipop(k):
    global operations_count,unit
    num_of_elements=0
    for i in range(k):
        if len(stack)!=0:
            pop()
            num_of_elements+=1
            operations_count+=1
    unit.append(num_of_elements)
    print(f"Stack after popping {k} elements is {stack} and the unit array is {unit}")
    return num_of_elements
arr=[5,7,9,2,6,1,8,3]
stack=[]
stack_len=8
unit=[]
operations_count=0
for i in arr:
    print(f"element {i} : - ")
    if i<=len(stack):
        print(f"since {i} is less than equal to {len(stack)}, therefore we multipop {i} element/s")
        multipop(i)
    push(i)
print(f"T(n)={sum(unit)} and num of operations is {operations_count}")
print(f"Time complexity O({sum(unit)//operations_count})")